### DISCUSSION TITLE
VSCode shows an error while importing document object from pyscript module

### DISCUSSION THREAD

**Discussion by picturecube** [pyscript/pyscript#2266] [Q&A]
The pyscript documentation says that your first line has to be `from pyscript import document` to make your python script interact with HTML elements. I did that in VSCode, but it says `Import "pyscript" could not be resolved Pylance (reportMissingImports)`. How do I resolve this matter?

**Comment by WebReflection** [comment]
> How do I resolve this matter?

I suggest providing a minimal example that shows the issue as that errors doesn't come from PyScript and it's not clear how you are using PyScript in general, how you bootstrap it, and so on, thanks.

**Comment by gregersn** [comment]
This is a two part problem. 

One is because there is no `pyscript` package installed in your Python environment that VSCode is using. This could be  resolved with `pip install pyscript`. But, that package intself, as a pip installable package does not contain any type hints for what will be exposed on the web. So it won't be resolved by that either. 

The issue here is that if you use an editor that wants/expects Python type hints, they are not really available when writing the scripts that will use the `pyscript` module.

**Reply by WebReflection** [reply]
also worth saying that no IDE right now is happy about Python written within *HTML* content, which is another reason to prefer using `src="./file.py"` instead as IDEs at least can be sure the rest of the Python code is working and properly highlighted hinted.

About PyScript though, it exports the entirety of the Web (`form pyscript import window`, as example) so it's still not clear how we can tell the IDE that such reference is a proxy for the Web API world and it's not clear how one would expect that to match any Python API or signature in general, as even methods are different between JS and Python.

**Reply by gregersn** [reply]
Creating a type stub file (.pyi) that would match how it is used in Python might be the route to take for that. 
It would probably not be a small job, but maybe it could be converted  from Typescript programatically.

**Reply by WebReflection** [reply]
yeah ... we'd love to have some contributor helping out because it's indeed not trivial + I've never done that before (I am the JS chap, not the Python one ;-)) so any hint or at least any starting point to check, improve, work on, consider, would be **very much welcomed** 🙏

**Reply by gregersn** [reply]
I might take a starting pass at it, however, there is a concern with where it should live, then. Currently, installing the pyscript package with pip will give the pyscript-cli package, with that having the typing stubs in that package would be the most straight forward and easiest. Alternatively that package maybe should be renamed to `pyscript-cli` on pypi, and a package named `pyscript` could only hold type stubs and supporting files for working with pyscript from editors.

But as a start and to see if it is useful at all to add some typestubs, I'll add it to the `pyscript-cli` package. It will work fine, and the cli tool is a supporting tool, so it still makes sense.

**Reply by gregersn** [reply]
A very quick, not very useful, proof of concept commit, that has not been tested in any way. There will be some issues with organizing files in the right way if it is to live next to pyscript-cli, maybe. 

https://github.com/gregersn/pyscript-cli/commit/81aa52cdbb449a3f944b7b19bf2660f92dc39698

**Reply by WebReflection** [reply]
@ntoll what are your thoughts around this?

**Reply by ntoll** [reply]
So this is a perennial topic for discussion. What @gregersn says is a great summary of the situation. I also appreciate @WebReflection's response since a type stub file is a non-trivial undertaking and maintenance burden. I also feel it's a bit of an impossible situation since PyScript gives you the whole of the web via the FFI API and so we're trying to side load something huge. :grimacing: 

And then there's the `pyscript` package on PyPI that (as pointed out) should perhaps be renamed to `pyscript-cli` or something more accurate.

Personally, if someone in the community wants to do the leg work for the type stubbing... go for it and we'll also try to support as best we can. However, this is a can of worms that behaves like Doctor Who's Tardis... it's a lot bigger inside the can of worms than it looks from the outside. :worm:

**Reply by gregersn** [reply]
If I am going to start using typescript, I will probably be writing stubs for myself, if none are available, add to it as I use things. I can then either keep that to myself, or make it available. 
Having some of it typed, like the upper imports, as "Any" is still better, to me, than having nothing. As explicitly saying that  something could be anything is better than it just being "I don't know what this is, at all"
And then expanding on it as  it is being worked on. Yes, it is a big undertaking, but at least the documentation on MDN is quite good. 

Making it just a stubs-package, might actually be the better option.  Then the current `pyscript` package can keep its name, or change it, but it is easier to just start working on it (I have, but not pushed it anywhere yet), and it is better to not mix the two, as the type of contributions to them is very different.

**Reply by ntoll** [reply]
@gregersn OK... how about a `pyscript-stubs` package..? Happy to create a repos under the `pyscript` org here on GitHub and add you to it as a collaborator if that would be useful. :+1: Happy to be led by what works for you. :hugs:

**Reply by gregersn** [reply]
@ntoll Sounds good.  Just let me get it started and have it kinda package ready first. And I'll get back to you.

**Reply by ntoll** [reply]
@gregersn perfect. Just ping me here (or on the PyScript discord) and I'll get things set up.

Also, I followed links to your homepage. I'm really a classically trained musician, so when I wonder into a place full of artistic coding I grin like the Cheshire Cat. :grin: This is something we want to encourage with PyScript. Expressive coding.

**Reply by WebReflection** [reply]
FWIWI I think if we have a *pyscript* module in pip that should reflect *pyscript* and not *pyscript-cli* ... but on the other hand, I find *pyscript-stubs* an explicit solution, although I still find current *pyscript* module confusing for new comers.

**Reply by gregersn** [reply]
I have finally gotten around to start working on a package for stubs. 
https://github.com/gregersn/pyscript-stubs

**Reply by Josverl** [reply]
There is also [Micropython-webassembly-stubs](https://pypi.org/project/micropython-webassembly-stubs/) on pypi , that has both pyscript stubs , and MicroPython.

Available for MicroPython 1.24.1 and preview.
https://github.com/Josverl/micropython-stubs/tree/main/publish%2Fmicropython-v1_25_0_preview-webassembly-stubs


Perhaps it would be good to see how these can be joined up ?

I maintain both the stubs, as well as the stubbier tool to merge type annotations between different stubs,
So would be happy to collaborate 
 

Simplest would be to add to the pyscript ['reference stub' ](https://github.com/Josverl/micropython-stubs/tree/main/reference%2Fpyscript) that I have generated through script , and manually updated from the various doc pages.

That could be used both as a stand-alone, but probably more useful with a complete MicroPython package as I now publish it.
But there may be use-cases I have not considered, so please let me know if I miss the point

**Reply by Josverl** [reply]
Also happy to move that reference part into the pyscript org, 
That is the reason I abstract the reference from the publication as I'm planning a similar move for the MicroPython reference stubs to the MicroPython repo.

**Reply by gregersn** [reply]
@Josverl I would be happy to help work on whatever, instead. I just really want to be able to have some typing while writing client side pyscript for webbrowsers. I do not feel strongly about my started project. I will look at the links you provided.

**Comment by ntoll** [comment]
It's wonderful to see such collaboration and cooperation. You mention bringing such efforts under the PyScript org... happy to help. What do you have in mind? :+1:

**Comment by Josverl** [comment]
I was thinking to use the same approach I have been working towards for MicroPython as a whole , and adapt it to fit the tools and usecase for PyScript

- Use a single source of truth for documentation, type checking and IDE hints
  This are `.pyi` type info stubs, including docstrings in .Rst or .MD formatting for modules , classes, methods etc. 
  We can bootstrap these based on the type-stubs that I have created ( or generate a more up-to-date version ) and add in the ones the @gregersn has started on.
  
- We can use tooling ( I use Sphinx_AutoApi for MicroPython) but there are similar tools for MkDocs to generate both complete  document pages , as well as documentation fragments that can be included in another documentation page. We'll need to look which tool does a good job, and fits with the tools you already use.
[💻 API documentation building](https://github.com/mkdocs/catalog?tab=readme-ov-file#-api-documentation-building)

- than means that the existing pages that contain documentation can be simplified by adding AutoAPI placeholders 
- if desired , a complete Module / API reference can be autogenerated.
(capabilities will depend on the tools used) 

**A sample PoC I built last year for MicroPython** 
Result : https://josverl.github.io/MP_AutoApi/ 

_Basic page template:_

- Source  https://github.com/Josverl/MP_AutoApi/blob/Sphinx_AutoAPI/docs/library/btree.rst
- Stub : https://github.com/Josverl/MP_AutoApi/tree/Sphinx_AutoAPI/docs/stubs/btree
- Result : https://josverl.github.io/MP_AutoApi/library/btree.html

_Document with AutoAPI fragments mixed in :_ 

- Source: https://github.com/Josverl/MP_AutoApi/blob/Sphinx_AutoAPI/docs/library/esp32.rst
- Stub: https://github.com/Josverl/MP_AutoApi/blob/Sphinx_AutoAPI/docs/stubs/esp32/
- Result :  https://josverl.github.io/MP_AutoApi/library/esp32.html


Then there are two workflows: 

### To update documentation :
1. Make an update to the stubs, Add / change method , or parameters , add an overload etc 
2. [A GH Action kicks off](https://github.com/Josverl/MP_AutoApi/actions/runs/14173986848) and builds updated  docs with the update ( optionally per version, filtered on branch / approvals etc ) 
3. Publish to GH pages / elsewhere 

###Stub publication:
There are multiple ways this could be done: 

**Publish a a stub-only package just for the pyscript module to PyPI** 
  This can be used both for Pyodide and MicroPython wasm/EmScripten ( sorry if I use the wrong terms ) 
  Important that this should be in pep-0561 stub only package. 
       Distribution : types-pyscript  ( IMO pyscript-stubs should be  widely accepted as well ) 
      contains the stub-only package : pyscrupt-stubs , with the following structure
      pyscript-stubs
      ├── __init__.pyi
      ├── foo.pyi
      └──  bar .pyi

Publication can be done the same tooling currently used to publish packages 
      
- Pyodide used could just install that package 
- for MicroPython I can configure the wasm/webassembly package to depend on the `types-pyscript`/`pyscript-stub` package 


### Location 

the location for the stub files should be somewhere that 
- Allows updates by maintainers , and contributors 
- Is , or can be made available (pulled in),  during build of the documentation 

so It can be either 
- in the pyscript source tree - simplest for the maintainers 
  such as https://github.com/pyscript/pyscript/tree/main/core/**stubs**/stdlib/pyscript
- in the documentation ( perhaps simplest for documenters and contributers  , and automation) 
  https://github.com/pyscript/docs/**stubs**/pyscript

- or in its own repo ( isolation ) 
   https://github.com/pyscript/**stubs**/pyscript

**Reply by gregersn** [reply]
My issue with basing the typing on the source, is that a lot of the class names are not matching the class names used in the MDN documentation, which makes it a lot  easier to actually look up and figure out how stuff works when writing the code.

**Reply by Josverl** [reply]
I assume you are referring to accessing the HTML DOM elements, is that correct ?
as I understand these originate from polyscript 
```py    
    window = polyscript.xworker.window
    document = window.document
    sync = polyscript.xworker.sync
```
This part I have not yet looked into - It should be possible to generate a basic stub for the polyscript module as well ( assuming there is some refection working) 
But pretty sure it would require manual updates.

On matching the class names between MDN / Python.
Should be simple to generate docpages that allow searching on either by using the templating and combining actual documentation and indexes , with the info from the docstrings.

Can you give an example snippet of code that you would like a stub for ?

**Comment by Josverl** [comment]
I think the maintainers should decide where the stubs should be placed and maintained,
 - Im happy to send a PR with my current state to whatever location

We also need to select a tool for mkdocs AutoAPI stub --> docs translations. 
There are a few candidates - for Sphinx I have tried out and compared a few different ones on capabilities , configuration and simplicity of use.  This can be dane with a simple Foo.Bar package , or the above 

Then we need to start doing the work 

perhaps we can discuss details during one of the calls to decide

**Reply by ntoll** [reply]
How about a new "stubs" repos under the PyScript org on GitHub?

**Reply by Josverl** [reply]
That would work, I think it would also be good to add some goals & guidelines on what to include, and how to handle any differences between CPython and MicroPython, license etc etc. but can build that as we go along

**Comment by ntoll** [comment]
OK. Since we needed a starting point, I have forked @gregersn's `pyscript-stubs` repository here: https://github.com/pyscript/pyscript-stubs

**Reply by Josverl** [reply]
Possibly a simpler way would have been to move the existing repo, as now there is an unclear fork. 

There is an option in repo >Settings to detach from the fork network ( Greger;s original repo).
![image](https://github.com/user-attachments/assets/35c89a8e-e949-4b33-8cdd-72af4703c996)
its a bit of a hassle now , but would make ownership clearer.

**Comment by ntoll** [comment]
I've created a PR that I'd like @gregersn to check: https://github.com/gregersn/pyscript-stubs/pull/1

It simply adds a README, LICENSE and CODE_OF_CONDUCT (as required by the PyScript project). Greg, please do let me know if there's a problem here and we can figure out next steps. Otherwise, please add a comment to the PR referenced above saying you're OK with these additions, and then we can start work. :+1:

**Reply by gregersn** [reply]
I just merged it, before I read here. Looks good

**Reply by Josverl** [reply]
a few points :

Im confused by the location of the license. 
- if the root repo is Gregers, who is the licenser ?

- my current work is MIT licensed, 
It would be silly for me to submit a major body of work, that uses a license that requires me to use your license rather than MIT
in short :

> Derivative works of the software will need to comply with the more detailed requirements of the Apache License 2.0, such as preserving NOTICE files and providing attribution for changes.

I need to think this over, I love to collaborate, and MIT and Apache 2.0 are not that different. The intents are the same.

But it would have a potential ripple effect on my own software the several thousands of downloads of the stubs each month , and the various typing tools that already use the micropython-stubs under a MIT license. Also it feels like signing over my work to someone else , without a proper dialog on how to do this.

I'd love to find a way to work together, and for me to keep using the MIT license for the parts that I create, and apply the other licenses just to the parts created and shared by others.

**Comment by ramalho** [comment]
On Wed, Feb 5, 2025 at 8:04 AM Greger ***@***.***> wrote:
> If I am going to start using typescript, I will probably be writing stubs for myself, if none are available, add to it as I use things. I can then either keep that to myself, or make it available.

That would be a cool contribution, Greger. Please share the stubs you
create, even if they are just a start. That's how Linus Torvalds built
Linux with the help of thousands!

> Having some of it typed, like the upper imports, as "Any" is still better, to me, than having nothing. As explicitly saying that something could be anything is better than it just being "I don't know what this is, at all"

I disagree with that approach. It's better to leave code without type
hints than annotating with a lot of `Any` hints without much
consideration.

The reason is that `Any` defeats type analysis, but its effect is not
limited to the place where it is used. A single `Any` type hint can
spread around the typing graph via type inference, and has far
reaching effects disabling type checking throughout the codebase.

If we don't know how (or don't have the resources) to annotate
something it's better to leave it without annotation, perhaps adding a
#type: ignore directive to silence the type checker.

> Making it just a stubs-package, might actually be the better option.

I agree!

Cheers,

Luciano



-- 
Luciano Ramalho
|  Author of Fluent Python (O'Reilly, 2015-2022)
|  https://www.oreilly.com/library/view/fluent-python-2nd/9781492056348/
|  Mastodon: @***@***.***

**Reply by Josverl** [reply]
The _typeshed module defines the `Incomplete` type specific for partial typing.

By using Incomplete instead of Any, it clarifies that the argument types still need work. It communicates that improvements or more specific types should be made in the future, making it easier to spot areas for enhancement in stubs 

https://github.com/python/typeshed/issues/5613
